File: Languages/Nim/web_server.nim
# Hello, World!
echo "Hello, World!"

import json, httpclient, os, times, strutils

type
  TestResult = object
    endpoint: string
    url: string
    server_hash: string
    expected_hash: string
    status: string
    error: string

  FinalResult = object
    total_tests: int
    passed: int
    failed: int
    success_rate: string
    timestamp: string
    results: seq[TestResult]

proc loadCompareJson(filename: string): JsonNode =
  try:
    let content = readFile(filename)
    parseJson(content)
  except:
    echo "Error: Could not read " & filename
    nil

proc makeHttpRequest(url: string): string =
  var client = newHttpClient(timeout = 5000)
  try:
    let response = client.get(url)
    if response.status == "200 OK":
      return response.body
    else:
      return ""
  except:
    return ""

proc testWebServer() =
  let baseUrl = "http://localhost:3000"
  let compareFile = "../webserver/compare.json"
  let outputFile = "test-result.json"
  
  # Load expected hashes
  let expectedHashes = loadCompareJson(compareFile)
  if expectedHashes == nil:
    return
  
  echo "Testing 100 endpoints..."
  
  var results: seq[TestResult]
  var passed = 0
  var failed = 0
  
  # Test each endpoint
  for i in 0..<100:
    let endpoint = "test-" & $i
    let url = baseUrl & "/" & endpoint
    
    let response = makeHttpRequest(url)
    
    var result = TestResult(
      endpoint: endpoint,
      url: url,
      expected_hash: expectedHashes{endpoint}.getStr(""),
      status: "",
      error: ""
    )
    
    if response == "":
      result.status = "FAILED"
      result.error = "HTTP request failed"
      inc(failed)
    else:
      try:
        let data = parseJson(response)
        let serverHash = data{"hash"}.getStr("")
        result.server_hash = serverHash
        
        if serverHash == result.expected_hash:
          result.status = "PASSED"
          inc(passed)
        else:
          result.status = "FAILED"
          inc(failed)
      except:
        result.status = "FAILED"
        result.error = "JSON parse error"
        inc(failed)
    
    results.add(result)
    
    # Progress indicator
    if (i + 1) mod 10 == 0:
      echo "Tested " & $(i + 1) & "/100 endpoints..."
  
  # Create final result
  let totalTests = results.len
  let successRate = formatFloat(passed.float / totalTests.float * 100, ffDecimal, 1) & "%"
  
  let finalResult = FinalResult(
    total_tests: totalTests,
    passed: passed,
    failed: failed,
    success_rate: successRate,
    timestamp: getTime().format("yyyy-MM-dd HH:mm:ss"),
    results: results
  )
  
  # Save results to file
  try:
    let output = pretty(finalResult, 4)
    writeFile(outputFile, output)
    echo "Results saved to: " & outputFile
  except:
    echo "Error saving results"
  
  echo "\nTest completed!"
  echo "Passed: " & $passed & "/" & $totalTests & " (" & successRate & ")"
  echo "Failed: " & $failed & "/" & $totalTests & " (" & formatFloat(failed.float / totalTests.float * 100, ffDecimal, 1) & "%)"

testWebServer()

File: Languages/Haskell/WebServer.hs
module Main where

import System.IO
import Network.HTTP.Client
import Data.Aeson
import Data.Aeson.Types
import Data.ByteString.Lazy (ByteString)
import qualified Data.ByteString.Lazy as BL
import qualified Data.ByteString.Char8 as BC
import Data.Time
import Control.Exception
import Control.Monad

-- Hello, World!
hello :: IO ()
hello = putStrLn "Hello, World!"

data TestResult = TestResult
    { endpoint :: String
    , url :: String
    , serverHash :: Maybe String
    , expectedHash :: String
    , status :: String
    , error :: Maybe String
    } deriving (Show, Eq)

instance ToJSON TestResult where
    toJSON (TestResult endpoint' url' serverHash' expectedHash' status' error') =
        object [ "endpoint" .= endpoint'
               , "url" .= url'
               , "server_hash" .= serverHash'
               , "expected_hash" .= expectedHash'
               , "status" .= status'
               , "error" .= error'
               ]

data FinalResult = FinalResult
    { totalTests :: Int
    , passed :: Int
    , failed :: Int
    , successRate :: String
    , timestamp :: String
    , results :: [TestResult]
    } deriving (Show, Eq)

instance ToJSON FinalResult where
    toJSON (FinalResult totalTests' passed' failed' successRate' timestamp' results') =
        object [ "total_tests" .= totalTests'
               , "passed" .= passed'
               , "failed" .= failed'
               , "success_rate" .= successRate'
               , "timestamp" .= timestamp'
               , "results" .= results'
               ]

loadCompareJSON :: FilePath -> IO (Maybe Object)
loadCompareJSON filename = do
    exists <- doesFileExist filename
    if not exists
        then do
            putStrLn $ "Error: Could not find " ++ filename
            return Nothing
        else do
            content <- BL.readFile filename
            case decode content of
                Nothing -> do
                    putStrLn "Error: Could not parse JSON"
                    return Nothing
                Just obj -> return $ Just obj

makeHTTPRequest :: Manager -> String -> IO (Maybe ByteString)
makeHTTPRequest manager url = do
    request <- parseRequest url
    catch (do
        response <- httpLbs request manager
        if statusCode (responseStatus response) == 200
            then return $ Just (responseBody response)
            else return Nothing
    ) (\(_ :: SomeException) -> return Nothing)

testWebServer :: IO ()
testWebServer = do
    let baseUrl = "http://localhost:3000"
    let compareFile = "../webserver/compare.json"
    
    -- Load expected hashes
    maybeExpected <- loadCompareJSON compareFile
    case maybeExpected of
        Nothing -> return ()
        Just expected -> do
            putStrLn "Testing 100 endpoints..."
            
            manager <- newManager defaultManagerSettings
            
            let testEndpoint i = do
                    let endpoint = "test-" ++ show i
                    let url = baseUrl ++ "/" ++ endpoint
                    
                    response <- makeHTTPRequest manager url
                    
                    let expectedHash = case lookup endpoint (Object expected) of
                                        Just (String h) -> h
                                        _ -> ""
                    
                    case response of
                        Just resp -> do
                            case decode resp of
                                Just (Object obj) -> do
                                    let serverHash = case lookup "hash" obj of
                                                        Just (String h) -> h
                                                        _ -> ""
                                    
                                    let status = if serverHash == expectedHash then "PASSED" else "FAILED"
                                    let passed = if status == "PASSED" then 1 else 0
                                    let failed = if status == "FAILED" then 1 else 0
                                    
                                    return (TestResult endpoint url (Just serverHash) expectedHash status Nothing, passed, failed)
                                _ -> do
                                    return (TestResult endpoint url Nothing expectedHash "FAILED" (Just "JSON parse error"), 0, 1)
                        Nothing -> do
                            return (TestResult endpoint url Nothing expectedHash "FAILED" (Just "HTTP request failed"), 0, 1)
            
            -- Test all endpoints
            results <- mapM testEndpoint [0..99]
            let testResults = map (\(r, _, _) -> r) results
            let totalPassed = sum $ map (\(_, p, _) -> p) results
            let totalFailed = sum $ map (\(_, _, f) -> f) results
            
            -- Progress indicator
            mapM_ (\i -> when ((i + 1) `mod` 10 == 0) $ 
                     putStrLn $ "Tested " ++ show (i + 1) ++ "/100 endpoints...") [0..99]
            
            -- Create final result
            let totalTests = length testResults
            let successRate = show (fromIntegral totalPassed / fromIntegral totalTests * 100 :: Double) ++ "%"
            
            currentTime <- getCurrentTime
            let timestamp = formatTime defaultTimeLocale "%Y-%m-%d %H:%M:%S" currentTime
            
            let finalResult = FinalResult totalTests totalPassed totalFailed successRate timestamp testResults
            
            -- Save results to file
            let outputFile = "test-result.json"
            BL.writeFile outputFile (encode finalResult)
            putStrLn $ "Results saved to: " ++ outputFile
            
            putStrLn "\nTest completed!"
            putStrLn $ "Passed: " ++ show totalPassed ++ "/" ++ show totalTests ++ " (" ++ successRate ++ ")"
            putStrLn $ "Failed: " ++ show totalFailed ++ "/" ++ show totalTests ++ " (" ++ show (100 - read (takeWhile (/= '%') successRate) :: Double) ++ "%)"

main :: IO ()
main = do
    hello
    testWebServer
